plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# OPTIONS_GHC -Wno-orphans       #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+module PlutusIR.Core.Instance.Pretty.Readable where
+
+import PlutusCore.Core.Instance.Pretty.Readable
+import PlutusCore.Default.Universe
+import PlutusCore.Pretty.ConfigName
+import PlutusCore.Pretty.PrettyConst
+import PlutusCore.Pretty.Readable
+import PlutusIR.Core.Type
+import PlutusPrelude
+import Prettyprinter
+
+
+prettyConfigReadable :: PrettyConfigReadable PrettyConfigName
+prettyConfigReadable = botPrettyConfigReadable defPrettyConfigName ShowKindsYes
+
+prettyReadable :: PrettyBy (PrettyConfigReadable PrettyConfigName) a => a -> Doc ann
+prettyReadable = prettyBy prettyConfigReadable
+
+viewApp :: Term tyname name uni fun ann
+        -> (Term tyname name uni fun ann, [Either (Type tyname uni ann) (Term tyname name uni fun ann)])
+viewApp t = go t []
+  where
+    go (Apply _ t s)  args = go t (Right s : args)
+    go (TyInst _ t a) args = go t (Left a : args)
+    go t args              = (t, args)
+
+vcatHard :: [Doc ann] -> Doc ann
+vcatHard = concatWith (\x y -> x <> hardline <> y)
+
+(<?>) :: Doc ann -> Doc ann -> Doc ann
+p <?> q = align . nest 2 $ sep [p, q]
+
+infixr 6 <?>
+
+type PrettyConstraints configName tyname name uni fun =
+  ( PrettyReadableBy configName tyname
+  , PrettyReadableBy configName name
+  , GShow uni, Closed uni, uni `Everywhere` PrettyConst, Pretty fun
+  )
+
+instance PrettyConstraints configName tyname name uni fun
+          => PrettyBy (PrettyConfigReadable configName) (Term tyname name uni fun a) where
+    prettyBy = inContextM $ \case
+        Constant _ con         -> unitDocM $ pretty con
+        Builtin _ bi           -> unitDocM $ pretty bi
+        t | appLike t          ->
+          compoundDocM juxtFixity $ \ prettyIn ->
+            let ppArg (Left a)  = braces $ prettyIn ToTheRight botFixity a
+                ppArg (Right t) = prettyIn ToTheRight juxtFixity t
+            in prettyIn ToTheLeft juxtFixity fun <?> fillSep (map ppArg args)
+          where
+            appLike Apply{}  = True
+            appLike TyInst{} = True
+            appLike _        = False
+
+            (fun, args)                  = view t []
+            view (Apply _ fun arg)  args = view fun (Right arg:args)
+            view (TyInst _ fun arg) args = view fun (Left arg:args)
+            view fun                args = (fun, args)
+        Apply{} -> error "The impossible happened. Get a lottery ticket."
+        TyInst{} -> error "The impossible happened. Get a lottery ticket."
```

Comment:
I agree this is impossible, but it doesn't hurt to be specific. "Impossible: this case should be covered by the 'appLike' guard"